# London Housing Case Study - Tier 3 Complete

## Objectives
Hey! Great job getting through those challenging DataCamp courses. You're learning a lot in a short span of time. 

In this notebook, you're going to apply the skills you've been learning, bridging the gap between the controlled environment of DataCamp and the *slightly* messier work that data scientists do with actual datasets!

Here’s the mystery we’re going to solve: ***which boroughs of London have seen the greatest increase in housing prices, on average, over the last two decades?***


A borough is just a fancy word for district. You may be familiar with the five boroughs of New York… well, there are 32 boroughs within Greater London [(here's some info for the curious)](https://en.wikipedia.org/wiki/London_boroughs). Some of them are more desirable areas to live in, and the data will reflect that with a greater rise in housing prices.

***This is the Tier 3 notebook, which means it's not filled in at all: we'll just give you the skeleton of a project, the brief and the data. It's up to you to play around with it and see what you can find out! Good luck! If you struggle, feel free to look at easier tiers for help; but try to dip in and out of them, as the more independent work you do, the better it is for your learning!***

This challenge will make use of only what you learned in the following DataCamp courses: 
- Prework courses (Introduction to Python for Data Science, Intermediate Python for Data Science)
- Data Types for Data Science
- Python Data Science Toolbox (Part One) 
- pandas Foundations
- Manipulating DataFrames with pandas
- Merging DataFrames with pandas

Of the tools, techniques and concepts in the above DataCamp courses, this challenge should require the application of the following: 
- **pandas**
    - **data ingestion and inspection** (pandas Foundations, Module One) 
    - **exploratory data analysis** (pandas Foundations, Module Two)
    - **tidying and cleaning** (Manipulating DataFrames with pandas, Module Three) 
    - **transforming DataFrames** (Manipulating DataFrames with pandas, Module One)
    - **subsetting DataFrames with lists** (Manipulating DataFrames with pandas, Module One) 
    - **filtering DataFrames** (Manipulating DataFrames with pandas, Module One) 
    - **grouping data** (Manipulating DataFrames with pandas, Module Four) 
    - **melting data** (Manipulating DataFrames with pandas, Module Three) 
    - **advanced indexing** (Manipulating DataFrames with pandas, Module Four) 
- **matplotlib** (Intermediate Python for Data Science, Module One)
- **fundamental data types** (Data Types for Data Science, Module One) 
- **dictionaries** (Intermediate Python for Data Science, Module Two)
- **handling dates and times** (Data Types for Data Science, Module Four)
- **function definition** (Python Data Science Toolbox - Part One, Module One)
- **default arguments, variable length, and scope** (Python Data Science Toolbox - Part One, Module Two) 
- **lambda functions and error handling** (Python Data Science Toolbox - Part One, Module Four) 

## The Data Science Pipeline

This is Tier Three, so we'll get you started. But after that, it's all in your hands! When you feel done with your investigations, look back over what you've accomplished, and prepare a quick presentation of your findings for the next mentor meeting. 

Data Science is magical. In this case study, you'll get to apply some complex machine learning algorithms. But as  [David Spiegelhalter](https://www.youtube.com/watch?v=oUs1uvsz0Ok) reminds us, there is no substitute for simply **taking a really, really good look at the data.** Sometimes, this is all we need to answer our question.

Data Science projects generally adhere to the four stages of Data Science Pipeline:
1. Sourcing and loading 
2. Cleaning, transforming, and visualizing 
3. Modeling 
4. Evaluating and concluding 


### 1. Sourcing and Loading 

Any Data Science project kicks off by importing  ***pandas***. The documentation of this wonderful library can be found [here](https://pandas.pydata.org/). As you've seen, pandas is conveniently connected to the [Numpy](http://www.numpy.org/) and [Matplotlib](https://matplotlib.org/) libraries. 

***Hint:*** This part of the data science pipeline will test those skills you acquired in the pandas Foundations course, Module One. 

#### 1.1. Importing Libraries

In [1]:
# Let's import the pandas, numpy libraries as pd, and np respectively. 
import pandas as pd

# Load the pyplot collection of functions from matplotlib, as plt 
from matplotlib import pyplot as plt

#### 1.2.  Loading the data
Your data comes from the [London Datastore](https://data.london.gov.uk/): a free, open-source data-sharing portal for London-oriented datasets. 

In [2]:
# First, make a variable called url_LondonHousePrices, and assign it the following link, enclosed in quotation-marks as a string:
# https://data.london.gov.uk/download/uk-house-price-index/70ac0766-8902-4eb5-aab5-01951aaed773/UK%20House%20price%20index.xls

url_LondonHousePrices = "https://data.london.gov.uk/download/uk-house-price-index/70ac0766-8902-4eb5-aab5-01951aaed773/UK%20House%20price%20index.xls"

# The dataset we're interested in contains the Average prices of the houses, and is actually on a particular sheet of the Excel file. 
# As a result, we need to specify the sheet name in the read_excel() method.
# Put this data into a variable called properties.  
properties = pd.read_excel(url_LondonHousePrices, sheet_name='Average price', index_col= None)
properties

Unnamed: 0 City of London Barking & Dagenham       Barnet       Bexley  \
0          NaT      E09000001          E09000002    E09000003    E09000004   
1   1995-01-01    91448.98487         50460.2266  93284.51832  64958.09036   
2   1995-02-01    82202.77314        51085.77983  93190.16963  64787.92069   
3   1995-03-01    79120.70256        51268.96956  92247.52435  64367.49344   
4   1995-04-01    77101.20804        53133.50526  90762.87492  64277.66881   
..         ...            ...                ...          ...          ...   
342 2023-06-01         982370             343972       570542       383969   
343 2023-07-01         960158             340763       575808       387475   
344 2023-08-01         888278             343731       586501       387360   
345 2023-09-01         875234             344255       589399       389590   
346 2023-10-01         948692             334460       597538       393227   

           Brent      Bromley       Camden      Croydon       Ealing  ...  \
0      E09000005    E09000006    E09000007    E09000008    E09000009  ...   
1    71306.56698  81671.47692  120932.8881  69158.16225  79885.89069  ...   
2    72022.26197  81657.55944  119508.8622  68951.09542  80897.06551  ...   
3    72015.76274  81449.31143  120282.2131  68712.44341  81379.86288  ...   
4    72965.63094  81124.41227   120097.899  68610.04641  82188.90498  ...   
..           ...          ...          ...          ...          ...  ...   
342       526985       496422       872180       402962       508547  ...   
343       525460       500377       845434       406623       509818  ...   
344       552411       504524       857874       413175       514537  ...   
345       554568       505185       830524       419400       515339  ...   
346       542864       512900       820541       401511       499337  ...   

      NORTH WEST YORKS & THE HUMBER EAST MIDLANDS WEST MIDLANDS  \
0      E12000002          E12000003     E12000004     E12000005   
1    43958.48001        44803.42878   45544.52227   48527.52339   
2    43925.42289        44528.80721   46051.57066   49341.29029   
3     44434.8681        45200.46775   45383.82395   49442.17973   
4     44267.7796        45614.34341   46124.23045   49455.93299   
..           ...                ...           ...           ...   
342       211797             206557        247251        248088   
343       213954             209645        248027        250741   
344       215859             213663        252436        254762   
345       216665             208529        248945        254351   
346       215719             208188        245632        253130   

    EAST OF ENGLAND       LONDON   SOUTH EAST   SOUTH WEST Unnamed: 47  \
0         E12000006    E12000007    E12000008    E12000009         NaN   
1        56701.5961  74435.76052  64018.87894   54705.1579         NaN   
2       56593.59475  72777.93709  63715.02399  54356.14843         NaN   
3       56171.18278  73896.84204  64113.60858  53583.07667         NaN   
4       56567.89582  74455.28754  64623.22395  54786.01938         NaN   
..              ...          ...          ...          ...         ...   
342          347186       519147       386845       321229         NaN   
343          350894       525318       390783       323504         NaN   
344          353977       532113       394212       329711         NaN   
345          353810       531777       392646       327388         NaN   
346          348615       515504       389223       330260         NaN   

         England  
0      E92000001  
1    53202.77128  
2     53096.1549  
3     53201.2843  
4     53590.8548  
..           ...  
342       302925  
343       305903  
344       310379  
345       308435  
346       305902  

[347 rows x 49 columns]

### 2. Cleaning, transforming, and visualizing
This second stage is arguably the most important part of any Data Science project. The first thing to do is take a proper look at the data. Cleaning forms the majority of this stage, and can be done both before or after Transformation.

The end goal of data cleaning is to have tidy data. When data is tidy: 

1. Each variable has a column.
2. Each observation forms a row.

Keep the end goal in mind as you move through this process, every step will take you closer. 



***Hint:*** This part of the data science pipeline should test those skills you acquired in: 
- Intermediate Python for data science, all modules.
- pandas Foundations, all modules. 
- Manipulating DataFrames with pandas, all modules.
- Data Types for Data Science, Module Four.
- Python Data Science Toolbox - Part One, all modules

**2.1. Exploring your data** 

Think about your pandas functions for checking out a dataframe. 

In [3]:
properties.head()

Unnamed: 0 City of London Barking & Dagenham       Barnet       Bexley  \
0        NaT      E09000001          E09000002    E09000003    E09000004   
1 1995-01-01    91448.98487         50460.2266  93284.51832  64958.09036   
2 1995-02-01    82202.77314        51085.77983  93190.16963  64787.92069   
3 1995-03-01    79120.70256        51268.96956  92247.52435  64367.49344   
4 1995-04-01    77101.20804        53133.50526  90762.87492  64277.66881   

         Brent      Bromley       Camden      Croydon       Ealing  ...  \
0    E09000005    E09000006    E09000007    E09000008    E09000009  ...   
1  71306.56698  81671.47692  120932.8881  69158.16225  79885.89069  ...   
2  72022.26197  81657.55944  119508.8622  68951.09542  80897.06551  ...   
3  72015.76274  81449.31143  120282.2131  68712.44341  81379.86288  ...   
4  72965.63094  81124.41227   120097.899  68610.04641  82188.90498  ...   

    NORTH WEST YORKS & THE HUMBER EAST MIDLANDS WEST MIDLANDS EAST OF ENGLAND  \
0    E12000002          E12000003     E12000004     E12000005       E12000006   
1  43958.48001        44803.42878   45544.52227   48527.52339      56701.5961   
2  43925.42289        44528.80721   46051.57066   49341.29029     56593.59475   
3   44434.8681        45200.46775   45383.82395   49442.17973     56171.18278   
4   44267.7796        45614.34341   46124.23045   49455.93299     56567.89582   

        LONDON   SOUTH EAST   SOUTH WEST Unnamed: 47      England  
0    E12000007    E12000008    E12000009         NaN    E92000001  
1  74435.76052  64018.87894   54705.1579         NaN  53202.77128  
2  72777.93709  63715.02399  54356.14843         NaN   53096.1549  
3  73896.84204  64113.60858  53583.07667         NaN   53201.2843  
4  74455.28754  64623.22395  54786.01938         NaN   53590.8548  

[5 rows x 49 columns]

**2.2. Cleaning the data**

You might find you need to transpose your dataframe, check out what its row indexes are, and reset the index. You  also might find you need to assign the values of the first row to your column headings  . (Hint: recall the .columns feature of DataFrames, as well as the iloc[] method).

Don't be afraid to use StackOverflow for help  with this.

In [4]:
properties = properties.transpose()

In [5]:
properties = properties.reset_index()
properties

index          0                    1                    2  \
0             Unnamed: 0        NaT  1995-01-01 00:00:00  1995-02-01 00:00:00   
1         City of London  E09000001          91448.98487          82202.77314   
2     Barking & Dagenham  E09000002           50460.2266          51085.77983   
3                 Barnet  E09000003          93284.51832          93190.16963   
4                 Bexley  E09000004          64958.09036          64787.92069   
5                  Brent  E09000005          71306.56698          72022.26197   
6                Bromley  E09000006          81671.47692          81657.55944   
7                 Camden  E09000007          120932.8881          119508.8622   
8                Croydon  E09000008          69158.16225          68951.09542   
9                 Ealing  E09000009          79885.89069          80897.06551   
10               Enfield  E09000010          72514.69096          73155.19746   
11             Greenwich  E09000011          62300.10169          60993.26863   
12               Hackney  E09000012          61296.52637          63187.08332   
13  Hammersmith & Fulham  E09000013          124902.8602           122087.718   
14              Haringey  E09000014          76287.56947          78901.21036   
15                Harrow  E09000015          84769.52599          83396.10525   
16              Havering  E09000016          68000.13774          69393.51294   
17            Hillingdon  E09000017          73834.82964           75031.0696   
18              Hounslow  E09000018          72231.70537          71051.55852   
19             Islington  E09000019          92516.48557          94342.37334   
20  Kensington & Chelsea  E09000020          182694.8326          182345.2463   
21  Kingston upon Thames  E09000021          80875.84843          81230.13524   
22               Lambeth  E09000022          67770.98843          65381.51908   
23              Lewisham  E09000023          60491.26109          60869.27091   
24                Merton  E09000024           82070.6133          79982.74872   
25                Newham  E09000025          53539.31919          53153.88306   
26             Redbridge  E09000026          72189.58437           72141.6261   
27  Richmond upon Thames  E09000027          109326.1245          111103.0394   
28             Southwark  E09000028          67885.20344           64799.0648   
29                Sutton  E09000029          71536.97357          70893.20851   
30         Tower Hamlets  E09000030          59865.18995          62318.53353   
31        Waltham Forest  E09000031          61319.44913          60252.12246   
32            Wandsworth  E09000032          88559.04381          88641.01678   
33           Westminster  E09000033          133025.2772          131468.3096   
34           Unnamed: 34        NaN                  NaN                  NaN   
35          Inner London  E13000001           78251.9765          75885.70201   
36          Outer London  E13000002          72958.79836          72937.88262   
37           Unnamed: 37        NaN                  NaN                  NaN   
38            NORTH EAST  E12000001          42076.35411          42571.98949   
39            NORTH WEST  E12000002          43958.48001          43925.42289   
40    YORKS & THE HUMBER  E12000003          44803.42878          44528.80721   
41         EAST MIDLANDS  E12000004          45544.52227          46051.57066   
42         WEST MIDLANDS  E12000005          48527.52339          49341.29029   
43       EAST OF ENGLAND  E12000006           56701.5961          56593.59475   
44                LONDON  E12000007          74435.76052          72777.93709   
45            SOUTH EAST  E12000008          64018.87894          63715.02399   
46            SOUTH WEST  E12000009           54705.1579          54356.14843   
47           Unnamed: 47        NaN                  NaN                  NaN   
48               England  E92000001          53202

In [6]:

properties.columns = properties.iloc[0]
properties = properties.drop(properties.index[0])
properties


0             Unnamed: 0        NaT 1995-01-01 00:00:00 1995-02-01 00:00:00  \
1         City of London  E09000001         91448.98487         82202.77314   
2     Barking & Dagenham  E09000002          50460.2266         51085.77983   
3                 Barnet  E09000003         93284.51832         93190.16963   
4                 Bexley  E09000004         64958.09036         64787.92069   
5                  Brent  E09000005         71306.56698         72022.26197   
6                Bromley  E09000006         81671.47692         81657.55944   
7                 Camden  E09000007         120932.8881         119508.8622   
8                Croydon  E09000008         69158.16225         68951.09542   
9                 Ealing  E09000009         79885.89069         80897.06551   
10               Enfield  E09000010         72514.69096         73155.19746   
11             Greenwich  E09000011         62300.10169         60993.26863   
12               Hackney  E09000012         61296.52637         63187.08332   
13  Hammersmith & Fulham  E09000013         124902.8602          122087.718   
14              Haringey  E09000014         76287.56947         78901.21036   
15                Harrow  E09000015         84769.52599         83396.10525   
16              Havering  E09000016         68000.13774         69393.51294   
17            Hillingdon  E09000017         73834.82964          75031.0696   
18              Hounslow  E09000018         72231.70537         71051.55852   
19             Islington  E09000019         92516.48557         94342.37334   
20  Kensington & Chelsea  E09000020         182694.8326         182345.2463   
21  Kingston upon Thames  E09000021         80875.84843         81230.13524   
22               Lambeth  E09000022         67770.98843         65381.51908   
23              Lewisham  E09000023         60491.26109         60869.27091   
24                Merton  E09000024          82070.6133         79982.74872   
25                Newham  E09000025         53539.31919         53153.88306   
26             Redbridge  E09000026         72189.58437          72141.6261   
27  Richmond upon Thames  E09000027         109326.1245         111103.0394   
28             Southwark  E09000028         67885.20344          64799.0648   
29                Sutton  E09000029         71536.97357         70893.20851   
30         Tower Hamlets  E09000030         59865.18995         62318.53353   
31        Waltham Forest  E09000031         61319.44913         60252.12246   
32            Wandsworth  E09000032         88559.04381         88641.01678   
33           Westminster  E09000033         133025.2772         131468.3096   
34           Unnamed: 34        NaN                 NaN                 NaN   
35          Inner London  E13000001          78251.9765         75885.70201   
36          Outer London  E13000002         72958.79836         72937.88262   
37           Unnamed: 37        NaN                 NaN                 NaN   
38            NORTH EAST  E12000001         42076.35411         42571.98949   
39            NORTH WEST  E12000002         43958.48001         43925.42289   
40    YORKS & THE HUMBER  E12000003         44803.42878         44528.80721   
41         EAST MIDLANDS  E12000004         45544.52227         46051.57066   
42         WEST MIDLANDS  E12000005         48527.52339         49341.29029   
43       EAST OF ENGLAND  E12000006          56701.5961         56593.59475   
44                LONDON  E12000007         74435.76052         72777.93709   
45            SOUTH EAST  E12000008         64018.87894         63715.02399   
46            SOUTH WEST  E12000009          54705.1579         54356.14843   
47           Unnamed: 47        NaN                 NaN                 NaN   
48               England  E92000001         53202.77128          53096.1549   

0  1995-03-01 00:00:00 1995-04-01 00:00:00 1995-05-01 00:00:00  \
1          79120.70256         77101.20804         84409.14932

In [7]:
properties = properties.reset_index(drop = True)
properties

0             Unnamed: 0        NaT 1995-01-01 00:00:00 1995-02-01 00:00:00  \
0         City of London  E09000001         91448.98487         82202.77314   
1     Barking & Dagenham  E09000002          50460.2266         51085.77983   
2                 Barnet  E09000003         93284.51832         93190.16963   
3                 Bexley  E09000004         64958.09036         64787.92069   
4                  Brent  E09000005         71306.56698         72022.26197   
5                Bromley  E09000006         81671.47692         81657.55944   
6                 Camden  E09000007         120932.8881         119508.8622   
7                Croydon  E09000008         69158.16225         68951.09542   
8                 Ealing  E09000009         79885.89069         80897.06551   
9                Enfield  E09000010         72514.69096         73155.19746   
10             Greenwich  E09000011         62300.10169         60993.26863   
11               Hackney  E09000012         61296.52637         63187.08332   
12  Hammersmith & Fulham  E09000013         124902.8602          122087.718   
13              Haringey  E09000014         76287.56947         78901.21036   
14                Harrow  E09000015         84769.52599         83396.10525   
15              Havering  E09000016         68000.13774         69393.51294   
16            Hillingdon  E09000017         73834.82964          75031.0696   
17              Hounslow  E09000018         72231.70537         71051.55852   
18             Islington  E09000019         92516.48557         94342.37334   
19  Kensington & Chelsea  E09000020         182694.8326         182345.2463   
20  Kingston upon Thames  E09000021         80875.84843         81230.13524   
21               Lambeth  E09000022         67770.98843         65381.51908   
22              Lewisham  E09000023         60491.26109         60869.27091   
23                Merton  E09000024          82070.6133         79982.74872   
24                Newham  E09000025         53539.31919         53153.88306   
25             Redbridge  E09000026         72189.58437          72141.6261   
26  Richmond upon Thames  E09000027         109326.1245         111103.0394   
27             Southwark  E09000028         67885.20344          64799.0648   
28                Sutton  E09000029         71536.97357         70893.20851   
29         Tower Hamlets  E09000030         59865.18995         62318.53353   
30        Waltham Forest  E09000031         61319.44913         60252.12246   
31            Wandsworth  E09000032         88559.04381         88641.01678   
32           Westminster  E09000033         133025.2772         131468.3096   
33           Unnamed: 34        NaN                 NaN                 NaN   
34          Inner London  E13000001          78251.9765         75885.70201   
35          Outer London  E13000002         72958.79836         72937.88262   
36           Unnamed: 37        NaN                 NaN                 NaN   
37            NORTH EAST  E12000001         42076.35411         42571.98949   
38            NORTH WEST  E12000002         43958.48001         43925.42289   
39    YORKS & THE HUMBER  E12000003         44803.42878         44528.80721   
40         EAST MIDLANDS  E12000004         45544.52227         46051.57066   
41         WEST MIDLANDS  E12000005         48527.52339         49341.29029   
42       EAST OF ENGLAND  E12000006          56701.5961         56593.59475   
43                LONDON  E12000007         74435.76052         72777.93709   
44            SOUTH EAST  E12000008         64018.87894         63715.02399   
45            SOUTH WEST  E12000009          54705.1579         54356.14843   
46           Unnamed: 47        NaN                 NaN                 NaN   
47               England  E92000001         53202.77128          53096.1549   

0  1995-03-01 00:00:00 1995-04-01 00:00:00 1995-05-01 00:00:00  \
0          79120.70256         77101.20804         84409.14932

**2.3. Cleaning the data (part 2)**

You might we have to **rename** a couple columns. How do you do this? The clue's pretty bold...

In [8]:
properties.columns = [str(col).split(' ')[0] for col in properties.columns]
properties.rename(columns={'Unnamed:': 'London Boroughs', 'NaT': 'ID'}, inplace=True)
properties

London Boroughs         ID   1995-01-01   1995-02-01   1995-03-01  \
0         City of London  E09000001  91448.98487  82202.77314  79120.70256   
1     Barking & Dagenham  E09000002   50460.2266  51085.77983  51268.96956   
2                 Barnet  E09000003  93284.51832  93190.16963  92247.52435   
3                 Bexley  E09000004  64958.09036  64787.92069  64367.49344   
4                  Brent  E09000005  71306.56698  72022.26197  72015.76274   
5                Bromley  E09000006  81671.47692  81657.55944  81449.31143   
6                 Camden  E09000007  120932.8881  119508.8622  120282.2131   
7                Croydon  E09000008  69158.16225  68951.09542  68712.44341   
8                 Ealing  E09000009  79885.89069  80897.06551  81379.86288   
9                Enfield  E09000010  72514.69096  73155.19746  72190.44144   
10             Greenwich  E09000011  62300.10169  60993.26863  61377.83464   
11               Hackney  E09000012  61296.52637  63187.08332  63593.29935   
12  Hammersmith & Fulham  E09000013  124902.8602   122087.718  120635.9467   
13              Haringey  E09000014  76287.56947  78901.21036  78521.94855   
14                Harrow  E09000015  84769.52599  83396.10525  83416.23759   
15              Havering  E09000016  68000.13774  69393.51294  69368.02407   
16            Hillingdon  E09000017  73834.82964   75031.0696  74188.66949   
17              Hounslow  E09000018  72231.70537  71051.55852  72097.99411   
18             Islington  E09000019  92516.48557  94342.37334  93465.86407   
19  Kensington & Chelsea  E09000020  182694.8326  182345.2463  182878.8231   
20  Kingston upon Thames  E09000021  80875.84843  81230.13524  81111.48848   
21               Lambeth  E09000022  67770.98843  65381.51908  66336.51868   
22              Lewisham  E09000023  60491.26109  60869.27091  60288.03002   
23                Merton  E09000024   82070.6133  79982.74872  80661.68279   
24                Newham  E09000025  53539.31919  53153.88306  53458.26393   
25             Redbridge  E09000026  72189.58437   72141.6261  72501.35502   
26  Richmond upon Thames  E09000027  109326.1245  111103.0394  107325.4742   
27             Southwark  E09000028  67885.20344   64799.0648  65763.29719   
28                Sutton  E09000029  71536.97357  70893.20851  70306.83844   
29         Tower Hamlets  E09000030  59865.18995  62318.53353  63938.67686   
30        Waltham Forest  E09000031  61319.44913  60252.12246  60871.08493   
31            Wandsworth  E09000032  88559.04381  88641.01678  87124.81523   
32           Westminster  E09000033  133025.2772  131468.3096  132260.3417   
33           Unnamed: 34        NaN          NaN          NaN          NaN   
34          Inner London  E13000001   78251.9765  75885.70201  76591.59947   
35          Outer London  E13000002  72958.79836  72937.88262  72714.53478   
36           Unnamed: 37        NaN          NaN          NaN          NaN   
37            NORTH EAST  E12000001  42076.35411  42571.98949  42369.72984   
38            NORTH WEST  E12000002  43958.48001  43925.42289   44434.8681   
39    YORKS & THE HUMBER  E12000003  44803.42878  44528.80721  45200.46775   
40         EAST MIDLANDS  E12000004  45544.52227  46051.57066  45383.82395   
41         WEST MIDLANDS  E12000005  48527.52339  49341.29029  49442.17973   
42       EAST OF ENGLAND  E12000006   56701.5961  56593.59475  56171.18278   
43                LONDON  E12000007  74435.76052  72777.93709  73896.84204   
44            SOUTH EAST  E12000008  64018.87894  63715.02399  64113.60858   
45            SOUTH WEST  E12000009   54705.1579  54356.14843  53583.07667   
46           Unnamed: 47        NaN          NaN          NaN          NaN   
47               England  E92000001  53202.77128   53096.1549   53201.2843   

     1995-04-01   1995-05-01   1995-06-01   1995-07-01   1995-08-01  ...  \
0   77101.20804  84409.14932  94900.51244  110128.0423  112329.4376  ...   
1   53133.50526  53042.24852  5370

**2.4.Transforming the data**

Remember what Wes McKinney said about tidy data? 

You might need to **melt** your DataFrame here. 

In [12]:
# Try this here: 
clean_properties = pd.melt(properties, id_vars= ['London Boroughs', 'ID'])
clean_properties.rename(columns={'variable': 'Date', 'value': 'Price'}, inplace=True)
clean_properties

London Boroughs         ID        Date        Price
0          City of London  E09000001  1995-01-01  91448.98487
1      Barking & Dagenham  E09000002  1995-01-01   50460.2266
2                  Barnet  E09000003  1995-01-01  93284.51832
3                  Bexley  E09000004  1995-01-01  64958.09036
4                   Brent  E09000005  1995-01-01  71306.56698
...                   ...        ...         ...          ...
16603              LONDON  E12000007  2023-10-01       515504
16604          SOUTH EAST  E12000008  2023-10-01       389223
16605          SOUTH WEST  E12000009  2023-10-01       330260
16606         Unnamed: 47        NaN  2023-10-01          NaN
16607             England  E92000001  2023-10-01       305902

[16608 rows x 4 columns]

Remember to make sure your column data types are all correct. Average prices, for example, should be floating point numbers... 

In [13]:
clean_properties.dtypes

London Boroughs    object
ID                 object
Date               object
Price              object
dtype: object

In [14]:
clean_properties['Price'] = clean_properties['Price'].astype(float)
clean_properties.dtypes

London Boroughs     object
ID                  object
Date                object
Price              float64
dtype: object

**2.5. Cleaning the data (part 3)**

Do we have an equal number of observations in the ID, Average Price, Month, and London Borough columns? Remember that there are only 32 London Boroughs. How many entries do you have in that column? 

Check out the contents of the London Borough column, and if you find null values, get rid of them however you see fit. 

In [15]:
clean_properties = clean_properties[~clean_properties['London Boroughs'].astype(str).str.startswith('Unnamed')]
clean_properties['London Boroughs'].unique()

array(['City of London', 'Barking & Dagenham', 'Barnet', 'Bexley',
       'Brent', 'Bromley', 'Camden', 'Croydon', 'Ealing', 'Enfield',
       'Greenwich', 'Hackney', 'Hammersmith & Fulham', 'Haringey',
       'Harrow', 'Havering', 'Hillingdon', 'Hounslow', 'Islington',
       'Kensington & Chelsea', 'Kingston upon Thames', 'Lambeth',
       'Lewisham', 'Merton', 'Newham', 'Redbridge',
       'Richmond upon Thames', 'Southwark', 'Sutton', 'Tower Hamlets',
       'Waltham Forest', 'Wandsworth', 'Westminster', 'Inner London',
       'Outer London', 'NORTH EAST', 'NORTH WEST', 'YORKS & THE HUMBER',
       'EAST MIDLANDS', 'WEST MIDLANDS', 'EAST OF ENGLAND', 'LONDON',
       'SOUTH EAST', 'SOUTH WEST', 'England'], dtype=object)

**2.6. Visualizing the data**

To visualize the data, why not subset on a particular London Borough? Maybe do a line plot of Month against Average Price?

In [16]:
City_of_London = clean_properties[clean_properties["London Boroughs"] == "City of London"]

City_of_London

London Boroughs         ID        Date         Price
0      City of London  E09000001  1995-01-01   91448.98487
48     City of London  E09000001  1995-02-01   82202.77314
96     City of London  E09000001  1995-03-01   79120.70256
144    City of London  E09000001  1995-04-01   77101.20804
192    City of London  E09000001  1995-05-01   84409.14932
...               ...        ...         ...           ...
16368  City of London  E09000001  2023-06-01  982370.00000
16416  City of London  E09000001  2023-07-01  960158.00000
16464  City of London  E09000001  2023-08-01  888278.00000
16512  City of London  E09000001  2023-09-01  875234.00000
16560  City of London  E09000001  2023-10-01  948692.00000

[346 rows x 4 columns]

In [ ]:
plt.plot(City_of_London["variable"], City_of_London["value"])
plt.show

To limit the number of data points you have, you might want to extract the year from every month value your *Month* column. 

To this end, you *could* apply a ***lambda function***. Your logic could work as follows:
1. look through the `Month` column
2. extract the year from each individual value in that column 
3. store that corresponding year as separate column. 

Whether you go ahead with this is up to you. Just so long as you answer our initial brief: which boroughs of London have seen the greatest house price increase, on average, over the past two decades? 

In [18]:
City_of_London['Year'] = City_of_London['Date'].apply(lambda x: pd.to_datetime(x).year)
City_of_London

C:\Users\Alejandro\AppData\Local\Temp\ipykernel_10848\48089501.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  City_of_London['Year'] = City_of_London['Date'].apply(lambda x: pd.to_datetime(x).year)


London Boroughs         ID        Date         Price  Year
0      City of London  E09000001  1995-01-01   91448.98487  1995
48     City of London  E09000001  1995-02-01   82202.77314  1995
96     City of London  E09000001  1995-03-01   79120.70256  1995
144    City of London  E09000001  1995-04-01   77101.20804  1995
192    City of London  E09000001  1995-05-01   84409.14932  1995
...               ...        ...         ...           ...   ...
16368  City of London  E09000001  2023-06-01  982370.00000  2023
16416  City of London  E09000001  2023-07-01  960158.00000  2023
16464  City of London  E09000001  2023-08-01  888278.00000  2023
16512  City of London  E09000001  2023-09-01  875234.00000  2023
16560  City of London  E09000001  2023-10-01  948692.00000  2023

[346 rows x 5 columns]

**3. Modeling**

Consider creating a function that will calculate a ratio of house prices, comparing the price of a house in 2018 to the price in 1998.

Consider calling this function create_price_ratio.

You'd want this function to:
1. Take a filter of dfg, specifically where this filter constrains the London_Borough, as an argument. For example, one admissible argument should be: dfg[dfg['London_Borough']=='Camden'].
2. Get the Average Price for that Borough, for the years 1998 and 2018.
4. Calculate the ratio of the Average Price for 1998 divided by the Average Price for 2018.
5. Return that ratio.

Once you've written this function, you ultimately want to use it to iterate through all the unique London_Boroughs and work out the ratio capturing the difference of house prices between 1998 and 2018.

Bear in mind: you don't have to write a function like this if you don't want to. If you can solve the brief otherwise, then great! 

***Hint***: This section should test the skills you acquired in:
- Python Data Science Toolbox - Part One, all modules

In [27]:
clean_properties['Year'] = clean_properties['Date'].apply(lambda x: pd.to_datetime(x).year)

def create_price_ratio(London_Borough):
    temp1 = London_Borough[London_Borough["Year"] == 1998]
    temp1 = temp1["Price"].mean()
    temp2 = London_Borough[London_Borough["Year"] == 2018]
    temp2 = temp2["Price"].mean()
    return temp1/temp2
create_price_ratio(clean_properties[clean_properties["London Boroughs"] == "Camden"])

C:\Users\Alejandro\AppData\Local\Temp\ipykernel_10848\3865886199.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_properties['Year'] = clean_properties['Date'].apply(lambda x: pd.to_datetime(x).year)


0.20261973503252542

### 4. Conclusion
What can you conclude? Type out your conclusion below. 

Look back at your notebook. Think about how you might summarize what you have done, and prepare a quick presentation on it to your mentor at your next meeting. 

We hope you enjoyed this practical project. It should have consolidated your data hygiene and pandas skills by looking at a real-world problem involving just the kind of dataset you might encounter as a budding data scientist. Congratulations, and looking forward to seeing you at the next step in the course! 

In [34]:
All_Boroughs = clean_properties['London Boroughs'].unique()
All_Boroughs
def create_price_ratio(London_Borough):
    London_Borough1 = clean_properties[clean_properties["London Boroughs"] == London_Borough]
    temp1 = London_Borough1[London_Borough1["Year"] == 1998]
    temp1 = temp1["Price"].mean()
    temp2 = London_Borough1[London_Borough1["Year"] == 2018]
    temp2 = temp2["Price"].mean()
    return temp1/temp2
create_price_ratio("Camden")
All_Boroughs_Ratio = {}
for borough in All_Boroughs:
    All_Boroughs_Ratio[borough] = create_price_ratio(borough)
All_Boroughs_Ratio


{'City of London': 0.18862157770244364,
 'Barking & Dagenham': 0.2042225623539368,
 'Barnet': 0.229452741207858,
 'Bexley': 0.23535076540630112,
 'Brent': 0.2043086864360114,
 'Bromley': 0.24421308489837307,
 'Camden': 0.20261973503252542,
 'Croydon': 0.23803288028014052,
 'Ealing': 0.23194048191708758,
 'Enfield': 0.23455064269011858,
 'Greenwich': 0.20995010893854218,
 'Hackney': 0.16133493530705734,
 'Hammersmith & Fulham': 0.24167443054605853,
 'Haringey': 0.19475619095546956,
 'Harrow': 0.24635417785626296,
 'Havering': 0.23120155787014757,
 'Hillingdon': 0.23807975835429931,
 'Hounslow': 0.25148317824115635,
 'Islington': 0.20643891170300288,
 'Kensington & Chelsea': 0.19675491852791566,
 'Kingston upon Thames': 0.23416190234282552,
 'Lambeth': 0.20170435486140822,
 'Lewisham': 0.1835124676472171,
 'Merton': 0.21091380604361798,
 'Newham': 0.18848754146121072,
 'Redbridge': 0.2236545053715767,
 'Richmond upon Thames': 0.24967779731157863,
 'Southwark': 0.18127484171283462,
 'Sutt

In [36]:
All_Boroughs_Ratio = sorted(All_Boroughs_Ratio.items(), key=lambda x: x[1], reverse=True)
All_Boroughs_Ratio

[('NORTH EAST', 0.3535967231925915),
 ('WEST MIDLANDS', 0.30199964293728065),
 ('YORKS & THE HUMBER', 0.2979679995370457),
 ('NORTH WEST', 0.297316807594284),
 ('EAST MIDLANDS', 0.27527471457895053),
 ('SOUTH WEST', 0.2634700981993535),
 ('England', 0.26243599006976337),
 ('SOUTH EAST', 0.2612065640720062),
 ('Hounslow', 0.25148317824115635),
 ('Richmond upon Thames', 0.24967779731157863),
 ('Harrow', 0.24635417785626296),
 ('Bromley', 0.24421308489837307),
 ('Sutton', 0.24280551426824518),
 ('Hammersmith & Fulham', 0.24167443054605853),
 ('EAST OF ENGLAND', 0.23998652920722383),
 ('Hillingdon', 0.23807975835429931),
 ('Croydon', 0.23803288028014052),
 ('Bexley', 0.23535076540630112),
 ('Enfield', 0.23455064269011858),
 ('Kingston upon Thames', 0.23416190234282552),
 ('Ealing', 0.23194048191708758),
 ('Havering', 0.23120155787014757),
 ('Barnet', 0.229452741207858),
 ('Outer London', 0.22629811224913093),
 ('Redbridge', 0.2236545053715767),
 ('Tower Hamlets', 0.2161367227623553),
 ('LO

Label a Google Doc [Your Name_London Answers]. Include answers to the following:

1. What did you find? Which borough is the most expensive? Any other interesting trends?
2. How did you arrive at your conclusion?
3. What were the main challenges you encountered? How did you overcome them? What could you not overcome?
4. Is there anything you’d like to investigate deeper?

1. From the exercise, we can see that 'Hackney' has gotten the most expensive from 1998 to 2018. Using the ratios we created, we revealed that 'Hackney' used to cost 16% the price it did in 2018.
2. I arrived at my conclusion by running the ratio function through all boroughs to compare the prices of 1998 to 2018.
3. The main challenge was understanding what the step was asking for, and remembering all the meticulous coding details. I overcame these challenges by googling and reading the documentation as well as using my notes.
4. I would love to investigate further by analyzing where spikes and pitfalls occured in the market. It could be interesting to see a heat map of the land that represented mean price with a slider that controls time.
